pkg/detector/detector.go

```diff
@@ -132,24 +133,30 @@ func (d *ResourceDetector) Start(ctx context.Context) error {
 	d.clusterPolicyReconcileWorker.Run(d.ConcurrentClusterPropagationPolicySyncs, d.stopCh)
 
 	// watch and enqueue PropagationPolicy changes.
-	propagationPolicyGVR := schema.GroupVersionResource{
-		Group:    policyv1alpha1.GroupVersion.Group,
-		Version:  policyv1alpha1.GroupVersion.Version,
-		Resource: policyv1alpha1.ResourcePluralPropagationPolicy,
-	}
 	policyHandler := fedinformer.NewHandlerOnEvents(d.OnPropagationPolicyAdd, d.OnPropagationPolicyUpdate, nil)
-	d.InformerManager.ForResource(propagationPolicyGVR, policyHandler)
-	d.propagationPolicyLister = d.InformerManager.Lister(propagationPolicyGVR)
+	ppInformer, err := d.ControllerRuntimeCache.GetInformer(ctx, &policyv1alpha1.PropagationPolicy{})
+	if err != nil {
+		klog.Errorf("Failed to get informer for PropagationPolicy: %v", err)
+		return err
+	}
+	_, err = ppInformer.AddEventHandler(policyHandler)
+	if err != nil {
+		klog.Errorf("Failed to add event handler for PropagationPolicy: %v", err)
+		return err
+	}
 
 	// watch and enqueue ClusterPropagationPolicy changes.
-	clusterPropagationPolicyGVR := schema.GroupVersionResource{
-		Group:    policyv1alpha1.GroupVersion.Group,
-		Version:  policyv1alpha1.GroupVersion.Version,
-		Resource: policyv1alpha1.ResourcePluralClusterPropagationPolicy,
-	}
 	clusterPolicyHandler := fedinformer.NewHandlerOnEvents(d.OnClusterPropagationPolicyAdd, d.OnClusterPropagationPolicyUpdate, nil)
-	d.InformerManager.ForResource(clusterPropagationPolicyGVR, clusterPolicyHandler)
-	d.clusterPropagationPolicyLister = d.InformerManager.Lister(clusterPropagationPolicyGVR)
+	cppInformer, err := d.ControllerRuntimeCache.GetInformer(ctx, &policyv1alpha1.ClusterPropagationPolicy{})
+	if err != nil {
+		klog.Errorf("Failed to get informer for ClusterPropagationPolicy: %v", err)
+		return err
+	}
+	_, err = cppInformer.AddEventHandler(clusterPolicyHandler)
+	if err != nil {
+		klog.Errorf("Failed to add event handler for ClusterPropagationPolicy: %v", err)
+		return err
+	}
```

Comment:
you need to put those two `AddEventHandler()` after the `d.Processor.Run(d.ConcurrentResourceTemplateSyncs, d.stopCh)`, because `AddEventHandler()` will immediately start the workers, and in the end of the workers logic, they will call `d.Processor.Add()`

Reply:
you mean d.Processor must run before d.ReconcilePropagationPolicy or d.ReconcileClusterPropagationPolicy trigger logic? If d.Processor has not been created, d.Processor.Add() might have problems?

Reply:
yes, occasional panic on startup

Reply:
I got it, tks

Reply:
> because AddEventHandler() will immediately start the workers,

Which workers we are talking about here?

Reply:
> Which workers we are talking about here?

`policyReconcileWorker` and `clusterPolicyReconcileWorker`. Before this commit, these workers would be started after the `InformerManager` was started, so there would be no problem.